api/python/quilt3/data_transfer.py

```diff
@@ -85,6 +85,72 @@ class S3ClientProvider:
 
     We assume that public buckets are read-only: write operations should always use S3ClientProvider.standard_client
     """
+    _event_callbacks = {}
+
+    @staticmethod
+    def add_options_safely(params: dict, options: dict):
+        """
+        Add options to the params dictionary safely.
+        This method ensures that the options do not overwrite existing keys in the params dictionary.
+        """
+        for key, value in options.items():
+            if key in params:
+                raise ValueError(f"Cannot override key `{key}` using options: {options}.")
+            params[key] = value
+
+    @classmethod
+    def reset_event_callbacks(cls):
+        cls._event_callbacks = {}
+
+    @classmethod
+    def register_event_options(cls, event_name: str, **kwargs: dict) -> None:
+        """Register options for S3 client events. A convenience method over register_event_callback
+        for trivial callbacks that always and only add options.
+
+        Args:
+            event_name: The name of the event to register for (e.g. 'creating-client-class')
```

Comment:
AFAICT `creating-client-class` doesn't make sense because it needs to be registered on session, see https://boto3.amazonaws.com/v1/documentation/api/latest/guide/events.html#creating-client-class
if we'd allow to set this handler on session I guess it could be used for client modification I mentioned including registration of `provide-client-params` event handlers